Src/IronPython.Modules/_ctypes/Field.cs

```
@@ -112,110 +100,138 @@ public sealed class Field : PythonTypeDataSlot, ICodeFormattable {
                 }
-                return value;
             }
 
             /// <summary>
-            /// Called for fields which have been limited to a range of bits.  Sets the 
+            /// Called for fields which have been limited to a range of bits.  Sets the
             /// specified value into the bits for the field.
             /// </summary>
             private void SetBitsValue(MemoryHolder address, int baseOffset, object value) {
                 // get the value in the form of a ulong which can contain the biggest bitfield
                 ulong newBits;
-                if (value is int) {
-                    newBits = (ulong)(int)value;
-                } else if (value is BigInteger) {
-                    newBits = (ulong)(long)(BigInteger)value;
+                if (IsBoolType) {
+                    newBits = PythonOps.IsTrue(value) ? 1ul : 0ul;
                 } else {
-                    throw PythonOps.TypeErrorForTypeMismatch("int or long", value);
+                    value = PythonOps.Index(value); // since 3.8
+                    if (value is int iVal) {
+                        newBits = (ulong)iVal;
+                    } else if (value is BigInteger biVal) {
+                        if (biVal.Sign >= 0) {
+                            if (biVal <= ulong.MaxValue) {
+                                newBits = (ulong)biVal;
+                            } else {
+                                newBits = (ulong)(biVal & ulong.MaxValue);
+                            }
+                        } else {
+                            if (biVal >= long.MinValue) {
+                                newBits = (ulong)(long)biVal;
+                            } else {
+                                newBits = (ulong)(biVal & ulong.MaxValue);
+                            }
+                        }
+                    } else {
+                        throw PythonOps.TypeErrorForTypeMismatch("int", value);
```

I think this bit is unreachable so you can throw InvalidOperationException?

---

Right. It's a left-over from a pre-3.8 code version.